File: location-tracker-user-service/src/main/java/org/saphka/locationtracker/user/config/SecurityConfiguration.java
package org.saphka.locationtracker.user.config;

import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.HttpMethod;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
import org.springframework.security.config.http.SessionCreationPolicy;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

@Configuration
@EnableWebSecurity
public class SecurityConfiguration extends WebSecurityConfigurerAdapter {

    @Value("${jwt.secret.file}")
    private String jwtFilePath;

    @Value("${jwt.issuer}")
    private String jwtIssuer;

    @Override
    protected void configure(HttpSecurity http) throws Exception {
        http.csrf().disable()
                .authorizeRequests()
                .antMatchers().permitAll()
                .antMatchers(HttpMethod.POST, "/users/register").permitAll()
                .anyRequest().authenticated()
                .and()
                .addFilter(new JWTAuthenticationFilter(authenticationManager(), "/users/auth", jwtSecret(), jwtIssuer))
                .addFilter(new JWTAuthorizationFilter(authenticationManager(), jwtSecret(), jwtIssuer))
                .sessionManagement().sessionCreationPolicy(SessionCreationPolicy.STATELESS);
    }

    @Bean
    public PasswordEncoder passwordEncoder() {
        return new BCryptPasswordEncoder();
    }

    @Bean
    @Qualifier("jwt-secret")
    public String jwtSecret() throws IOException {
        return Files.readString(Path.of(jwtFilePath));
    }

}

File: src/main/kotlin/org/saphka/location/tracker/user/configuration/SecurityConfiguration.kt
package org.saphka.location.tracker.user.configuration

import org.saphka.location.tracker.user.configuration.properties.JwtProperties
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder
import java.security.KeyPair
import java.security.KeyStore
import java.security.PrivateKey

private const val TOKEN_ALIAS = "token"
private const val BEARER = "Bearer "

@Configuration
class SecurityConfiguration {

    @Bean
    fun jwtKeyPair(jwtProperties: JwtProperties): KeyPair {
        with(KeyStore.getInstance("JKS")) {
            jwtProperties.keystore.inputStream.use {
                load(it, jwtProperties.keystorePassword.toCharArray())
            }
            val private = getKey(TOKEN_ALIAS, jwtProperties.keystorePassword.toCharArray()) as PrivateKey
            val public = getCertificate(TOKEN_ALIAS).publicKey

            return KeyPair(public, private)
        }
    }

    @Bean
    fun passwordEncoder(): PasswordEncoder = BCryptPasswordEncoder()

}